postgres/backup-to-s3.py

```diff
@@ -43,82 +44,94 @@
-            if not len(chunk):
-                print('Finished! Wrote {} chunks; {}'.format(
-                    chunks_done,
-                    humanize.naturalsize(chunks_done * CHUNK_SIZE)
-                ))
+
+class Backup(Thread):
+
+    def __init__(self, app_code_):
+        """
+        Args:
+            app_code_ (str): `kc` or `kpi`
+        """
+        self.__app_code = app_code_
+        super().__init__()
+
+    def run(self):
+        """
+        Backup postgres database for specific `app_code`.
+        """
+
+        s3connection = S3Connection(AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY)
+        s3bucket = s3connection.get_bucket(AWS_BUCKET)
+
+        DBDATESTAMP = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
+
+        # `postgis://` isn't recognized by `pg_dump`; replace it with `postgres://`
+        DBURL = re.sub(r'^postgis://', 'postgres://', APP_CODES.get(self.__app_code))
+        # Because we are running `pg_dump` within the container,
+        # we need to replace the hostname ...
+        DBURL = DBURL.replace(os.getenv("POSTGRES_HOST"), "127.0.0.1")
+        # ... and the port for '127.0.0.1:5432'
+        DBURL = re.sub(r"\:(\d+)\/", ":5432/", DBURL)
+
+        DUMPFILE = 'postgres-{}-{}-{}-{}.pg_dump'.format(
+            self.__app_code,
+            os.environ.get('PG_MAJOR'),
+            os.environ.get('PUBLIC_DOMAIN_NAME'),
+            DBDATESTAMP,
+        )
+
+        BACKUP_COMMAND = 'pg_dump --format=c --dbname="{}"'.format(DBURL)
+
+        # Determine where to put this backup
+        now = datetime.datetime.now()
+        for directory in DIRECTORIES:
+            prefix = directory['name'] + '/'
+            earliest_current_date = now - datetime.timedelta(days=directory['days'])
+            s3keys = s3bucket.list(prefix=prefix)
+            large_enough_backups = filter(lambda x: x.size >= MINIMUM_SIZE, s3keys)
+            young_enough_backup_found = False
+            for backup in large_enough_backups:
+                if parse_ts(backup.last_modified) >= earliest_current_date:
+                    young_enough_backup_found = True
+            if not young_enough_backup_found:
+                # This directory doesn't have any current backups; stop here and use it
+                # as the destination
                 break
-            s3backup.write(chunk)
-            chunks_done += 1
-            if '--hush' not in sys.argv:
-                print('Wrote {} chunks; {}'.format(
-                    chunks_done,
-                    humanize.naturalsize(chunks_done * CHUNK_SIZE)
-                ))
 
-    print('Backup `{}` successfully sent to S3.'.format(filename))
+        # Perform the backup
+        filename = ''.join((prefix, DUMPFILE))
+        print('Backing up to "{}"...'.format(filename))
+        upload = s3bucket.new_key(filename)
+        chunks_done = 0
+        with smart_open.smart_open(upload, 'wb') as s3backup:
+            process = subprocess.Popen(
```

Comment:
did you intend to lower the priority of `pg_dump` here or does that not matter for S3 backups?

Reply:
According to my tests, `pg_dump` receives the priority of its parent (i.e. `/kobo-docker-scripts/backup-to-s3.py`) so there is no needs to prefix `pg_dump` over here.

Reply:
Whoops! Right. :+1: